src/include/processor/operator/persistent/index_builder.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include <variant>
+
+#include "common/copy_constructors.h"
+#include "common/mpsc_queue.h"
+#include "common/static_vector.h"
+#include "common/types/internal_id_t.h"
+#include "common/types/types.h"
+#include "processor/execution_context.h"
+#include "storage/index/hash_index_builder.h"
+#include "storage/store/column_chunk.h"
+
+namespace kuzu {
+namespace processor {
+
+constexpr size_t BUFFER_SIZE = 1024;
+using IntBuffer = common::StaticVector<std::pair<int64_t, common::offset_t>, BUFFER_SIZE>;
+using StringBuffer = common::StaticVector<std::pair<std::string, common::offset_t>, BUFFER_SIZE>;
+
+class IndexBuilderGlobalQueues {
+public:
+    explicit IndexBuilderGlobalQueues(std::unique_ptr<storage::PrimaryKeyIndexBuilder> pkIndex);
+
+    void flushToDisk() const;
+
+    void insert(size_t index, StringBuffer elem);
+    void insert(size_t index, IntBuffer elem);
+
+    void consume();
+
+    common::LogicalTypeID pkTypeID() const { return pkIndex->keyTypeID(); }
+
+private:
+    void maybeConsumeIndex(size_t index);
+
+    std::array<std::mutex, storage::NUM_HASH_INDEXES> mutexes;
+    std::unique_ptr<storage::PrimaryKeyIndexBuilder> pkIndex;
+
+    using StringQueues = std::array<common::MPSCQueue<StringBuffer>, storage::NUM_HASH_INDEXES>;
+    using IntQueues = std::array<common::MPSCQueue<IntBuffer>, storage::NUM_HASH_INDEXES>;
+
+    // Queues for distributing primary keys.
+    std::variant<StringQueues, IntQueues> queues;
+};
+
+class IndexBuilderLocalBuffers {
+public:
+    explicit IndexBuilderLocalBuffers(IndexBuilderGlobalQueues& globalQueues);
+
+    void insert(std::string key, common::offset_t value);
+    void insert(int64_t key, common::offset_t value);
+
+    void flush();
+
+private:
+    IndexBuilderGlobalQueues* globalQueues;
+
+    // These arrays are much too large to be inline.
+    using StringBuffers = std::array<StringBuffer, storage::NUM_HASH_INDEXES>;
+    using IntBuffers = std::array<IntBuffer, storage::NUM_HASH_INDEXES>;
+    std::unique_ptr<StringBuffers> stringBuffers;
+    std::unique_ptr<IntBuffers> intBuffers;
+};
+
+class IndexBuilderSharedState {
+    friend class IndexBuilder;
+
+public:
+    explicit IndexBuilderSharedState(std::unique_ptr<storage::PrimaryKeyIndexBuilder> pkIndex);
+    void consume();
+    void flush() { globalQueues.flushToDisk(); }
+
+    void addProducer();
+    void quitProducer();
+    bool isDone() { return done.load(std::memory_order_relaxed); }
+
+private:
+    IndexBuilderGlobalQueues globalQueues;
+
+    std::atomic<size_t> producers;
+    std::atomic<bool> done;
+};
+
+// RAII for producer counting.
+class ProducerToken {
+public:
+    explicit ProducerToken(std::shared_ptr<IndexBuilderSharedState> sharedState)
+        : sharedState(std::move(sharedState)) {
+        this->sharedState->addProducer();
+    }
+    NO_COPY(ProducerToken);
+
+    void quit() {
```

Comment:
mark as inline. we took the convention to mark function implementations as `inline`,  though we missed some places I guess.

Reply:
From https://en.cppreference.com/w/cpp/language/inline:
> A function defined entirely inside a [class/struct/union definition](https://en.cppreference.com/w/cpp/language/classes), whether it's a member function or a non-member friend function, is implicitly an inline function unless it is attached to a [named module](https://en.cppreference.com/w/cpp/language/modules#Module_declarations)(since C++20).

Do we still want to inline it?

Reply:
I see. I'm fine with not explicitly inline it.